File: problems/leetcode/00088/go/solution.go
// Created by shiyang07ca at 2026/09/11 02:17
// leetgo: 1.4.17
// https://leetcode.cn/problems/merge-sorted-array/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func merge(nums1 []int, m int, nums2 []int, n int) {
	p1, p2, c := m-1, n-1, m+n-1
	for p2 >= 0 {
		if p1 >= 0 && nums1[p1] > nums2[p2] {
			nums1[c] = nums1[p1]
			p1--
		} else {
			nums1[c] = nums2[p2]
			p2--
		}
		c--
	}
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums1 := Deserialize[[]int](ReadLine(stdin))
	m := Deserialize[int](ReadLine(stdin))
	nums2 := Deserialize[[]int](ReadLine(stdin))
	n := Deserialize[int](ReadLine(stdin))
	merge(nums1, m, nums2, n)
	ans := nums1

	fmt.Println("\noutput:", Serialize(ans))
}

File: problems/leetcode/00088/python/solution.py
from __future__ import annotations

# Created by shiyang07ca at 2023/08/13 00:35
# leetgo: dev
# https://leetcode.cn/problems/merge-sorted-array/
from typing import cast

# leetgo_py does not ship type stubs.
# pyright: reportMissingTypeStubs=false
from leetgo_py import deserialize, read_line, serialize

# @lc code=begin


class Solution:
    def merge1(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """

        cur = m + n - 1
        while cur >= 0 and m > 0 and n > 0:
            if nums1[m - 1] >= nums2[n - 1]:
                nums1[cur] = nums1[m - 1]
                m -= 1
            else:
                nums1[cur] = nums2[n - 1]
                n -= 1
            cur -= 1
        while (cur >= 0 and m > 0) or n > 0:
            if m > 0:
                nums1[cur] = nums1[m - 1]
                m -= 1
            elif n > 0:
                nums1[cur] = nums2[n - 1]
                n -= 1
            cur -= 1

    # https://leetcode.cn/problems/merge-sorted-array/solutions/126371/88-by-ikaruga/
    def merge(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        i = len(nums1) - 1
        m, n = m - 1, n - 1
        while n >= 0:
            while m >= 0 and nums1[m] > nums2[n]:
                nums1[i], nums1[m] = nums1[m], nums1[i]
                i -= 1
                m -= 1
            nums1[i], nums2[n] = nums2[n], nums1[i]
            i -= 1
            n -= 1


# @lc code=end

if __name__ == "__main__":
    nums1 = cast(list[int], deserialize("List[int]", read_line()))
    m = cast(int, deserialize("int", read_line()))
    nums2 = cast(list[int], deserialize("List[int]", read_line()))
    n = cast(int, deserialize("int", read_line()))
    Solution().merge(nums1, m, nums2, n)
    ans = nums1

    print("\noutput:", serialize(ans))
